src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1287 @@
+           compile time */
+        CORRADE_INTERNAL_ASSERT(info->type != UFBX_OPEN_FILE_GEOMETRY_CACHE);
+
+        /* If we don't have a callback just defer to ufbx file loading */
+        if(!_callback) return ufbx_open_file(stream, path, path_len);
+
+        std::string file{path, path_len};
+        const Containers::Optional<Containers::ArrayView<const char>> data = _callback(file, InputFileCallbackPolicy::LoadTemporary, _userData);
+        if(!data) return false;
+
+        ufbx_open_memory_opts opts = {};
+        opts.allocator.allocator = info->temp_allocator;
+
+        /* We don't need to copy the file data as it's guaranteed to live for
+           the duration of the load function we are currently executing */
+        opts.no_copy = true;
+
+        return ufbx_open_memory(stream, data->data(), data->size(), &opts, NULL);
+    }
+
+    Containers::Optional<Containers::ArrayView<const char>> (*_callback)(const std::string&, InputFileCallbackPolicy, void*);
+    void* _userData;
+};
+
+struct MeshChunk {
+    /* ufbx_scene::meshes[] */
+    UnsignedInt meshId;
+    /* ufbx_mesh::materials[] (NOT ufbx_scene::materials[] !) */
+    UnsignedInt meshMaterialIndex;
+};
+
+struct FileTexture {
+    /* ufbx_scene::textures[] */
+    UnsignedInt textureIndex;
+    /* ufbx_scene::file_textures[] */
+    UnsignedInt fileTextureIndex;
+};
+
+}
+
+struct UfbxImporter::State {
+    ufbx_scene_ref scene;
+
+    /* Meshes split by material */
+    Containers::Array<MeshChunk> meshChunks;
+
+    /* Mapping from ufbx_scene::meshes[] -> State::meshChunks[] */
+    Containers::Array<UnsignedInt> meshChunkBase;
+
+    /* Offset subtracted from ufbx IDs to UfbxImporter::object() IDs, usually
+       one as the root node is excluded */
+    UnsignedInt nodeIdOffset = 0;
+    UnsignedInt objectCount = 0;
+
+    /* true if loaded from openFile(), false from openData() */
+    bool fromFile = false;
+
+    /* Name to ufbx_scene::texture_files[] */
+    std::unordered_map<std::string, UnsignedInt> imageNameMap;
+
+    /* Textures that have actual files */
+    Containers::Array<FileTexture> textures;
+
+    /* ufbx_scene::textures[] to State::textures[] */
+    Containers::Array<Int> textureRemap;
+
+    /* Cached AnyImageImporter for image2D() and image2DLevelCount */
+    UnsignedInt imageImporterId = ~UnsignedInt{};
+    Containers::Optional<AnyImageImporter> imageImporter;
+};
+
+UfbxImporter::UfbxImporter(PluginManager::AbstractManager& manager, const Containers::StringView& plugin): AbstractImporter{manager, plugin} {
+}
+
+UfbxImporter::~UfbxImporter() = default;
+
+ImporterFeatures UfbxImporter::doFeatures() const { return ImporterFeature::OpenData|ImporterFeature::FileCallback; }
+
+bool UfbxImporter::doIsOpened() const { return !!_state; }
+
+void UfbxImporter::doClose() { _state = nullptr; }
+
+void UfbxImporter::doOpenData(Containers::Array<char>&& data, const DataFlags) {
+    _state.reset();
```

Comment:
You don't need this -- the public `openData()` API calls `doOpenData()` only after the previous file (if any) is closed, and `doClose()` resets the `_state` already.

Same in `doOpenFile()` below.